app/src/main/java/co/tinode/tindroid/db/SqlStore.java

```diff
@@ -133,10 +134,10 @@ public boolean topicDelete(Topic topic) {
     }
 
     @Override
-    public Range getCachedMessagesRange(Topic topic) {
+    public MsgRange getCachedMessagesRange(Topic topic) {
```

Comment:
I also feel like this should be an array of ranges, not just one range.
The scenario is
* you subscribe to the topic, fetch `N` last messages starting from `L1` -> cached range is `[L1, L1 + N)`
* let some time pass and, say, in this your peer(s) posted `>>N` messages to the topic
* you connect to the topic again, fetch last `N` messages starting from `L2` -> now the sqlitedb in reality contains two ranges: `[L1, L1 + N)` and `[L2, L2 + N)`.
The problem is a single range won't be able to represent it. Either StoredTopic needs to store the range or we should reconstruct it on the fly from the message db every time we subscribe to the topic.

Reply:
Yes, the current logic of getting the bounds with `getCachedMessagesRange` is inadequate. The lower/upper bounds must the the least/greatest `seq`  not already present in a message and not contained within a deletion range.

> you subscribe to the topic, fetch N last messages starting from L1 -> cached range is [L1, L1 + N)

The current logic is to fetch 'last N messages'. The problem is slightly different: 

Say, sqlite already contains the range [Last, Last-N]. Client fetches older messages by asking the server `get count=N before=Last-N` which works fine if no messages were deleted. The problem starts if there is a deletion range [Del, Del + M], where Del+M < Last - N. The current logic breaks because the client will fetch `get count=N before=Del` and messages in the range [Del+N, Last-N] will never be retrieved. 

One option is to always fetch `get count=N before=Last-N` regardless of deletion ranges. It's wasteful. but it will produce correct outcome.  A nicer way is to understand that there are no more messages to fetch because all earlier messages were deleted. 

I'm going to implement "always `get count=N before=Last-N`", and the optimization can be done later.

Reply:
Sounds good to me.